pczt/src/sapling.rs

```diff
@@ -0,0 +1,548 @@
+                    (Ordering::Less, Ordering::Greater) | (Ordering::Greater, Ordering::Less) => {
+                        return None
+                    }
+                    // These cases mean that at least one of the two value sums is correct
+                    // and we can use it directly.
+                    (spends, outputs) => (
+                        matches!(spends, Ordering::Greater),
+                        matches!(outputs, Ordering::Greater),
+                    ),
+                };
+
+                if other_has_more_spends {
+                    self.spends.extend(spends.drain(self.spends.len()..));
+                }
+                if other_has_more_outputs {
+                    self.outputs.extend(outputs.drain(self.outputs.len()..));
+                }
+                if other_has_more_spends || other_has_more_outputs {
+                    // We check below that the overlapping spends and outputs match.
+                    // Assuming here that they will, we take the other bundle's value sum.
+                    self.value_sum = value_sum;
+                }
+            }
+        }
+
+        if self.anchor != anchor {
+            return None;
+        }
+
+        // Leverage the early-exit behaviour of zip to confirm that the remaining data in
+        // the other bundle matches this one.
+        for (lhs, rhs) in self.spends.iter_mut().zip(spends.into_iter()) {
+            // Destructure `rhs` to ensure we handle everything.
+            let Spend {
+                cv,
+                nullifier,
+                rk,
+                zkproof,
+                spend_auth_sig,
+                recipient,
+                value,
+                rcm,
+                rseed,
+                rcv,
+                proof_generation_key,
+                witness,
+                alpha,
+                zip32_derivation,
+                dummy_ask,
+                proprietary,
+            } = rhs;
+
+            if lhs.cv != cv || lhs.nullifier != nullifier || lhs.rk != rk {
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.zkproof, zkproof)
+                && merge_optional(&mut lhs.spend_auth_sig, spend_auth_sig)
+                && merge_optional(&mut lhs.recipient, recipient)
+                && merge_optional(&mut lhs.value, value)
+                && merge_optional(&mut lhs.rcm, rcm)
+                && merge_optional(&mut lhs.rseed, rseed)
+                && merge_optional(&mut lhs.rcv, rcv)
+                && merge_optional(&mut lhs.proof_generation_key, proof_generation_key)
+                && merge_optional(&mut lhs.witness, witness)
+                && merge_optional(&mut lhs.alpha, alpha)
+                && merge_optional(&mut lhs.zip32_derivation, zip32_derivation)
+                && merge_optional(&mut lhs.dummy_ask, dummy_ask)
+                && merge_map(&mut lhs.proprietary, proprietary))
```

Comment:
Here we can assume that if any of the fields on the LHS were `None`, then they were not needed to compute any sighash, and so there is no possibility of a merge invalidating a signature. It might be useful to point that out in a comment.